### PR TITLE
Snap on mesh elements

### PR DESCRIPTION
**Pull request by vcloarec**
Add a method in API to snap on mesh elements. 
The method returns the position of the snapped point on the closest mesh element :
- For vertex, the snapped position is the vertex position
- For edge, the snapped position is the projected point on the edge, extremity of edge if outside the edge
- For face, the snapped position is the centroïd of the face

The returned position is in map coordinates

### CODE REVIEW

File: src/core/mesh/qgsmeshlayer.cpp [hunk 1]
```diff
@@ -354,33 +354,15 @@ QgsMesh3dDataBlock QgsMeshLayer::dataset3dValue( const QgsMeshDatasetIndex &inde
 QgsMeshDatasetValue QgsMeshLayer::dataset1dValue( const QgsMeshDatasetIndex &index, const QgsPointXY &point, double searchRadius ) const
 {
   QgsMeshDatasetValue value;
+  QgsPointXY projectedPoint;
+  int selectedIndex = closestEdge( point, searchRadius, projectedPoint );
   QgsRectangle searchRectangle( point.x() - searchRadius, point.y() - searchRadius, point.x() + searchRadius, point.y() + searchRadius );
   const QgsTriangularMesh *mesh = triangularMesh();
   if ( mesh &&
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:361]: do we need this condition here if it is already in the `closestEdge` function?

File: src/core/mesh/qgsmeshlayer.cpp [hunk 2]
```diff
@@ -354,33 +354,15 @@ QgsMesh3dDataBlock QgsMeshLayer::dataset3dValue( const QgsMeshDatasetIndex &inde
 QgsMeshDatasetValue QgsMeshLayer::dataset1dValue( const QgsMeshDatasetIndex &index, const QgsPointXY &point, double searchRadius ) const
 {
   QgsMeshDatasetValue value;
+  QgsPointXY projectedPoint;
+  int selectedIndex = closestEdge( point, searchRadius, projectedPoint );
   QgsRectangle searchRectangle( point.x() - searchRadius, point.y() - searchRadius, point.x() + searchRadius, point.y() + searchRadius );
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:359]: is searchRectangle used in the function?

File: src/core/mesh/qgsmeshlayer.cpp [hunk 3]
```diff
@@ -484,6 +497,110 @@ void QgsMeshLayer::setReferenceTime( const QDateTime &referenceTime )
...
 
+QgsPointXY QgsMeshLayer::snapOnVertex( const QgsPointXY &point, double searchRadius )
+{
+  const QgsTriangularMesh *mesh = triangularMesh();
+  QgsPointXY exactPosition;
+  if ( !mesh )
+    return exactPosition;
+  QgsRectangle rectangle( point.x() - searchRadius, point.y() - searchRadius, point.x() + searchRadius, point.y() + searchRadius );
+  double maxDistance = searchRadius;
+  if ( mesh )
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:508]: already tested few lines above

File: src/core/mesh/qgsmeshlayer.cpp [hunk 4]
```diff
@@ -484,6 +497,110 @@ void QgsMeshLayer::setReferenceTime( const QDateTime &referenceTime )
...
+        exactPosition = vertex1;
+      }
+      if ( dist2 < maxDistance )
+      {
+        maxDistance = dist2;
+        exactPosition = vertex2;
+      }
+    }
+
+
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:531]: extra newline

File: src/core/mesh/qgsmeshlayer.h [hunk 5]
```diff
@@ -377,6 +377,23 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
       */
     void setReferenceTime( const QDateTime &referenceTime );
 
+    /**
+      * Returns the position of the snapped point on the closest mesh element
+      *
+      * For vertex, the snapped position is the vertex position
+      * For edge, the snapped position is the projected point on the edge, extremity of edge if outside the edge
+      * For face, the snapped position is the centroïd of the face
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h:385]: centroïd -> centroid

File: src/core/mesh/qgsmeshlayer.h [hunk 6]
```diff
@@ -377,6 +377,23 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
       */
     void setReferenceTime( const QDateTime &referenceTime );
 
+    /**
+      * Returns the position of the snapped point on the closest mesh element
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h:381]: if I am reading the code correctly, when there is an error (e.g. triangular mesh not intialized), it return invalid Point. good to add to the docs .

Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h:381]: closest mesh element .... add .... within search radius to docs

File: src/core/mesh/qgsmeshlayer.h [hunk 7]
```diff
@@ -467,6 +484,17 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
 
     QgsMeshDatasetIndex mStaticScalarDatasetIndex;
     QgsMeshDatasetIndex mStaticVectorDatasetIndex;
+
+    int closestEdge( const QgsPointXY &point, double searchRadius, QgsPointXY &projectedPoint ) const;
+
+    //! Returns the exact position in map coordinates of the closest vertex in the search area
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h:490]: do not need docs on private functions

**Review by PeterPetrik — APPROVED** [review]
good work!

File: src/core/mesh/qgsmeshlayer.cpp (hunk 2 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshlayer.cpp:359]: indeed, not anymore...

File: src/core/mesh/qgsmeshlayer.h (hunk 7 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshlayer.h:490]: not needed but convenient for later or other...

File: src/core/mesh/qgsmeshlayer.cpp (hunk 1 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshlayer.cpp:361]: no, indeed

**Review by PeterPetrik — APPROVED** [review]

**Comment by saberraz** [comment]
Gif for changelog
snap

**Pull request merged**